# Exploratory Data Analysis

## Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [2]:
df = pd.read_csv('flight.csv')

In [3]:
def preprocessing(df):
    """"
    Objective :
    Function for preprocessing data, known missing, duplicated values and basic stastics every column in dataframe and duplicated row
    df is dataframe
    """
    try:
        import pandas as pd
        variables = pd.DataFrame(columns=['Variable','Number of unique values','Percent of Null(%)','Number of Null','Type','Values'])
        for i, var in enumerate(df.columns):
            variables.loc[i] = [var, df[var].nunique(),df[var].isnull().sum()/df.shape[0]*100,df[var].isnull().sum(),df[var].dtypes,df[var].unique()]
        return (variables.set_index('Variable'))
    except:
        print('ada kesalahan penginputan')

## Checking Duplicated Data

In [11]:
df.duplicated().sum()

0

In [12]:
df.duplicated(subset='MEMBER_NO').sum()

0

## Checking Null Values

In [4]:
preprocessing(df)

Number of unique values  Percent of Null(%) Number of Null  \
Variable                                                                       
MEMBER_NO                           62988            0.000000              0   
FFP_DATE                             3068            0.000000              0   
FIRST_FLIGHT_DATE                    3406            0.000000              0   
GENDER                                  2            0.004763              3   
FFP_TIER                                3            0.000000              0   
WORK_CITY                            3234            3.602273           2269   
WORK_PROVINCE                        1165            5.156538           3248   
WORK_COUNTRY                          118            0.041278             26   
AGE                                    84            0.666794            420   
LOAD_TIME                               1            0.000000              0   
FLIGHT_COUNT                          153            0.000000              0   
BP_SUM                              23449            0.000000              0   
SUM_YR_1                            15828            0.874770            551   
SUM_YR_2                            16767            0.219089            138   
SEG_KM_SUM                          29081            0.000000              0   
LAST_FLIGHT_DATE                      731            0.000000              0   
LAST_TO_END                           731            0.000000              0   
AVG_INTERVAL                        10706            0.000000              0   
MAX_INTERVAL                          706            0.000000              0   
EXCHANGE_COUNT                         28            0.000000              0   
avg_discount                        54179            0.000000              0   
Points_Sum                          25062            0.000000              0   
Point_NotFlight                        99            0.000000              0   

                      Type                                             Values  
Variable                                                                       
MEMBER_NO            int64  [54993, 28065, 55106, 21189, 39546, 56972, 449...  
FFP_DATE            object  [11/2/2006, 2/19/2007, 2/1/2007, 8/22/2008, 4/...  
FIRST_FLIGHT_DATE   object  [12/24/2008, 8/3/2007, 8/30/2007, 8/23/2008, 4...  
GENDER              object                                [Male, Female, nan]  
FFP_TIER             int64                                          [6, 5, 4]  
WORK_CITY           object  [., nan, Los Angeles, guiyang, guangzhou, wulu...  
WORK_PROVINCE       object  [beijing, CA, guizhou, guangdong, xinjiang, zh...  
WORK_COUNTRY        object  [CN, US, FR, AN, JP, HK, MY, AU, NL, MX, CA, K...  
AGE                float64  [31.0, 42.0, 40.0, 64.0, 48.0, 46.0, 50.0, 43....  
LOAD_TIME           object                                        [3/31/2014]  
FLIGHT_COUNT         int64  [210, 140, 135, 23, 152, 92, 101, 73, 56, 64, ...  
BP_SUM               int64  [505308, 362480, 351159, 337314, 273844, 31333...  
SUM_YR_1           float64  [239560.0, 171483.0, 163618.0, 116350.0, 12456...  
SUM_YR_2           float64  [234188.0, 167434.0, 164982.0, 125500.0, 13070...  
SEG_KM_SUM           int64  [580717, 293678, 283712, 281336, 309928, 29458...  
LAST_FLIGHT_DATE    object  [3/31/2014, 3/25/2014, 3/21/2014, 12/26/2013, ...  
LAST_TO_END          int64  [1, 7, 11, 97, 5, 79, 3, 6, 15, 22, 67, 2, 65,...  
AVG_INTERVAL       float64  [3.483253589, 5.194244604, 5.298507463, 27.863...  
MAX_INTERVAL         int64  [18, 17, 73, 47, 52, 28, 45, 94, 95, 42, 112, ...  
EXCHANGE_COUNT       int64  [34, 29, 20, 11, 27, 10, 7, 5, 13, 1, 15, 3, 1...  
avg_discount       float64  [0.961639043, 1.25231444, 1.254675516, 1.09086...  
Points_Sum           int64  [619760, 415768, 406361, 372204, 338813, 34312...  
Point_NotFlight      int64  [50, 33, 26, 12, 39, 15, 29, 14, 7, 16, 3, 66,...

**Data Insight**

    1. GENDER, WORK_CITY, WORK_PROVINCE, WORK_COUNTRY, AGE, SUM_YR_1,SUM_YR_2 mempunyai null value
    2. Data sudah memiliki tipe data yang tepat

## Descriptitve Analysis

In [13]:
df.describe()

MEMBER_NO      FFP_TIER           AGE  FLIGHT_COUNT         BP_SUM  \
count  62988.000000  62988.000000  62568.000000  62988.000000   62988.000000   
mean   31494.500000      4.102162     42.476346     11.839414   10925.081254   
std    18183.213715      0.373856      9.885915     14.049471   16339.486151   
min        1.000000      4.000000      6.000000      2.000000       0.000000   
25%    15747.750000      4.000000     35.000000      3.000000    2518.000000   
50%    31494.500000      4.000000     41.000000      7.000000    5700.000000   
75%    47241.250000      4.000000     48.000000     15.000000   12831.000000   
max    62988.000000      6.000000    110.000000    213.000000  505308.000000   

            SUM_YR_1       SUM_YR_2     SEG_KM_SUM   LAST_TO_END  \
count   62437.000000   62850.000000   62988.000000  62988.000000   
mean     5355.376064    5604.026014   17123.878691    176.120102   
std      8109.450147    8703.364247   20960.844623    183.822223   
min         0.000000       0.000000     368.000000      1.000000   
25%      1003.000000     780.000000    4747.000000     29.000000   
50%      2800.000000    2773.000000    9994.000000    108.000000   
75%      6574.000000    6845.750000   21271.250000    268.000000   
max    239560.000000  234188.000000  580717.000000    731.000000   

       AVG_INTERVAL  MAX_INTERVAL  EXCHANGE_COUNT  avg_discount   Points_Sum  \
count  62988.000000  62988.000000    62988.000000  62988.000000   62988.0000   
mean      67.749788    166.033895        0.319775      0.721558   12545.7771   
std       77.517866    123.397180        1.136004      0.185427   20507.8167   
min        0.000000      0.000000        0.000000      0.000000       0.0000   
25%       23.370370     79.000000        0.000000      0.611997    2775.0000   
50%       44.666667    143.000000        0.000000      0.711856    6328.5000   
75%       82.000000    228.000000        0.000000      0.809476   14302.5000   
max      728.000000    728.000000       46.000000      1.500000  985572.0000   

       Point_NotFlight  
count     62988.000000  
mean          2.728155  
std           7.364164  
min           0.000000  
25%           0.000000  
50%           0.000000  
75%           1.000000  
max         140.000000

**Data Insight**

    1. nilai max AGE ada yang mencapai 110 tahun
    2. Total data yang dimiliki adalah 62988